File: urbanairship-core/src/main/java/com/urbanairship/analytics/SearchEventTemplate.java
/* Copyright Airship and Contributors */

package com.urbanairship.analytics;

import java.math.BigDecimal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * A class that represents a custom search event template for the application.
 */
public class SearchEventTemplate {

    /**
     * The search event template type.
     */
    @NonNull
    private static final String SEARCH_EVENT_TEMPLATE = "search";

    /**
     * The search event name.
     */
    @NonNull
    private static final String SEARCH_EVENT = "search";

    /**
     * The lifetime value property.
     */
    private static final String LIFETIME_VALUE = "ltv";

    /**
     * The type property.
     */
    private static final String TYPE = "type";

    /**
     * The query property.
     */
    private static final String QUERY = "query";

    /**
     * The category property.
     */
    private static final String CATEGORY = "category";

    /**
     * The ID property.
     */
    private static final String ID = "id";

    /**
     * The total results property.
     */
    private static final String TOTAL_RESULTS = "total_results";

    @NonNull
    private final String eventName;

    @Nullable
    private BigDecimal value;

    @Nullable
    private String type;

    @Nullable
    private String query;

    @Nullable
    private String category;

    @Nullable
    private String id;

    private long total_results;

    private SearchEventTemplate (@NonNull String eventName) {
        this.eventName = eventName;
    }

    public static SearchEventTemplate newSearchTemplate() {
        return new SearchEventTemplate(SEARCH_EVENT);
    }

    /**
     * Sets the event value.
     * <p>
     * The event's value will be accurate 6 digits after the decimal. The number must fall in the
     * range [-2^31, 2^31-1]. Any value outside that range will cause the event to be invalid.
     *
     * @param value The event's value as a BigDecimal.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setValue(@Nullable BigDecimal value) {
        this.value = value;
        return this;
    }

    /**
     * Sets the event value.
     * <p>
     * The event's value will be accurate 6 digits after the decimal. The number must fall in the
     * range [-2^31, 2^31-1]. Any value outside that range will cause the event to be invalid.
     *
     * @param value The event's value as a double. Must be a number.
     * @return An SearchEventTemplate.
     * @throws NumberFormatException if the value is infinity or not a number.
     */
    @NonNull
    public SearchEventTemplate setValue(double value) {
        return setValue(BigDecimal.valueOf(value));
    }

    /**
     * Sets the event value.
     * <p>
     * The event's value will be accurate 6 digits after the decimal. The number must fall in the
     * range [-2^31, 2^31-1]. Any value outside that range will cause the event to be invalid.
     *
     * @param value The event's value as a string. Must contain valid string representation of a big decimal.
     * @return An SearchEventTemplate.
     * @throws NumberFormatException if the event value does not contain a valid string representation
     * of a big decimal.
     */
    @NonNull
    public SearchEventTemplate setValue(@Nullable String value) {
        if (value == null || value.length() == 0) {
            this.value = null;
            return this;
        }

        return setValue(new BigDecimal(value));
    }

    /**
     * Sets the event value.
     * <p>
     * The event's value will be accurate 6 digits after the decimal. The number must fall in the
     * range [-2^31, 2^31-1]. Any value outside that range will cause the event to be invalid.
     *
     * @param value The event's value as an int.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setValue(int value) {
        return setValue(new BigDecimal(value));
    }

    /**
     * Set the type.
     *
     * @param type The type as a string.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setType(String type) {
        this.type = type;
        return this;
    }

    /**
     * Set the query.
     *
     * @param query The query as a string.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setQuery(String query) {
        this.query = query;
        return this;
    }

    /**
     * Set the category.
     *
     * @param category The category as a string.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setCategory(String category) {
        this.category = category;
        return this;
    }

    /**
     * Set the ID.
     *
     * @param id The ID as a string.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setId(String id) {
        this.id = id;
        return this;
    }

    /**
     * Set the total results.
     *
     * @param totalResults The total results as a string.
     * @return An SearchEventTemplate.
     */
    @NonNull
    public SearchEventTemplate setTotalResults(long totalResults) {
        this.total_results = totalResults;
        return this;
    }

    /**
     * Creates the custom search event.
     *
     * @return The custom search event.
     */
    @NonNull
    public CustomEvent createEvent() {
        CustomEvent.Builder builder = CustomEvent.newBuilder(this.eventName);

        if (this.value != null) {
            builder.setEventValue(this.value);
            builder.addProperty(LIFETIME_VALUE, true);
        } else {
            builder.addProperty(LIFETIME_VALUE, false);
        }

        if (this.type != null) {
            builder.addProperty(TYPE, this.type);
        }

        if (this.query != null) {
            builder.addProperty(QUERY, this.query);
        }

        if (this.category != null) {
            builder.addProperty(CATEGORY, this.category);
        }

        if (this.id != null) {
            builder.addProperty(ID, this.id);
        }

        builder.addProperty(TOTAL_RESULTS, this.total_results);

        builder.setTemplateType(SEARCH_EVENT_TEMPLATE);

        return builder.build();
    }
}

File: urbanairship-core/src/main/java/com/urbanairship/analytics/templates/SearchEventTemplate.kt
/* Copyright Airship and Contributors */

package com.urbanairship.analytics.templates

import com.urbanairship.json.JsonSerializable
import com.urbanairship.json.JsonValue
import com.urbanairship.json.jsonMapOf

/**
 * Search event template. Use with [com.urbanairship.analytics.CustomEvent.newBuilder]
 * or [com.urbanairship.analytics.customEvent] to create a custom event with search template properties.
 *
 * Example (Kotlin):
 * ```
 * customEvent(
 *     SearchEventTemplate.Type.SEARCH,
 *     SearchEventTemplate.Properties(
 *         category = "hotels",
 *         query = "beach resort",
 *         totalResults = 53
 *     )
 * ) {
 *     setEventValue(1.0)
 * }.track()
 * ```
 *
 * Example (Java):
 * ```
 * SearchEventTemplate.Properties properties = SearchEventTemplate.Properties.newBuilder()
 *     .setCategory("hotels")
 *     .setQuery("beach resort")
 *     .setTotalResults(53)
 *     .build();
 *
 * CustomEvent.newBuilder(SearchEventTemplate.Type.SEARCH, properties)
 *     .setEventValue(1.0)
 *     .build()
 *     .track();
 * ```
 *
 * @see com.urbanairship.analytics.CustomEvent.newBuilder
 * @see com.urbanairship.analytics.customEvent
 */
public sealed class SearchEventTemplate {

    public enum class Type(internal val eventName: String) {
        SEARCH("search")
    }

    public class Properties(
        /** The event's ID. */
        public val id: String? = null,

        /** The search query. */
        public val query: String? = null,

        /** The total search results */
        public val totalResults: Long? = null,

        /** The event's category. */
        public val category: String? = null,

        /** The event's type. */
        public val type: String? = null,

        /** If the value is a lifetime value or not. */
        public val isLTV: Boolean = false
    ): JsonSerializable {

        public companion object {
            private const val ID = "id"
            private const val QUERY = "query"
            private const val TOTAL_RESULTS = "total_results"
            private const val CATEGORY = "category"
            private const val TYPE = "type"
            private const val IS_LTV = "ltv"

            @JvmStatic
            public fun newBuilder(): Builder = Builder()
        }

        public class Builder(
            private var id: String? = null,
            private var query: String? = null,
            private var totalResults: Long? = null,
            private var category: String? = null,
            private var type: String? = null,
            private var isLTV: Boolean = false
        ) {

            /**
             * Set the type.
             *
             * @param type The type as a string.
             * @return An [Builder].
             */
            public fun setType(type: String?): Builder {
                return this.also { it.type = type }
            }

            /**
             * Set the query.
             *
             * @param query The query as a string.
             * @return An [Builder].
             */
            public fun setQuery(query: String?): Builder {
                return this.also { it.query = query }
            }

            /**
             * Set the category.
             *
             * @param category The category as a string.
             * @return An [Builder].
             */
            public fun setCategory(category: String?): Builder {
                return this.also { it.category = category }
            }

            /**
             * Set the ID.
             *
             * @param id The ID as a string.
             * @return An [Builder].
             */
            public fun setId(id: String?): Builder {
                return this.also { it.id = id }
            }

            /**
             * Set the total results.
             *
             * @param totalResults The total results as a string.
             * @return An [Builder].
             */
            public fun setTotalResults(totalResults: Long): Builder {
                return this.also { it.totalResults = totalResults }
            }

            /**
             * Set the life time value.
             *
             * @param value is life time value
             * @return A [Builder].
             */
            public fun setIsLifetimeValue(value: Boolean): Builder {
                return this.also { it.isLTV = value }
            }

            public fun build(): Properties = Properties(
                id = this.id,
                query = this.query,
                totalResults = this.totalResults,
                category = this.category,
                type = this.type,
                isLTV = this.isLTV
            )
        }

        override fun toJsonValue(): JsonValue = jsonMapOf(
            ID to id,
            QUERY to query,
            TOTAL_RESULTS to totalResults,
            CATEGORY to category,
            TYPE to type,
            IS_LTV to isLTV
        ).toJsonValue()
    }

    internal companion object {
        const val TEMPLATE_NAME = "search"
    }
}

/**
 * Creates a new [SearchEventTemplate.Properties].
 *
 * @param block A lambda function that configures the `Builder`.
 * @return A new `SearchEventTemplate.Properties` instance.
 */
@JvmSynthetic
public fun searchEventProperties(
    block: SearchEventTemplate.Properties.Builder.() -> Unit
): SearchEventTemplate.Properties {
    val builder = SearchEventTemplate.Properties.newBuilder()
    builder.block()
    return builder.build()
}
